server/architecture.md

```diff
@@ -0,0 +1,77 @@
+# Layers
+
+Most of the source code in this repository is split into tree layers.
+These layers are the repository layer, the service layer, and the API layer.
+There are a couple of advantages to maintain these three layers.
+For example, the repository layer decouples the business logic from database details.
+Similarly, the business logic in the service layer is independent of the API layer, e.g. a single service can be used for Graphql or Rest endpoints.
+This architecture is very similar to the so-called "clean architecture" in domain driven design.
+
+The three layers should not have any circular dependencies.
+For example, the repository layer can be used by the service layer or the API layer while the service layer can only be used by the API layer.
+This is enforced by putting the layers in separate Rust crates which can't have circular dependencies either.
+
+## Repository Layer
+
+The repository layer provides an abstraction over the used database.
+For example, entities returned from the repository layer look the same for Postgres and Sqlite.
+This database abstraction should be as small as possible, e.g. instead of adding specific queries for specific use-cases, existing query method should be reused where possible.
+
+Furthermore, the repository layer keeps the service logic decoupled from database schema changes.
+For example, a table schema change should only affect the repository that uses the changed table.
+This can be achieved by mapping the data returned from the database to a "domain object" which is database independent.
+For example, rows from the mSupply tables `name` and `item` are not referenced directly by their `id` but joined via a link table to allow name and item merges.
+While the database row may look like:
+
+```rust
```

Comment:
The following section is more how it should be rather than how it is implemented (from my experience with column rename for the merge changes). Let me know what you think about it.